scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordered_serialization/ProductLike.scala

```diff
@@ -0,0 +1,186 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+package com.twitter.scalding.macros.impl.ordered_serialization
+
+import scala.language.experimental.macros
+import scala.reflect.macros.Context
+
+import com.twitter.scalding._
+
+object ProductLike {
+  def compareBinary(c: Context)(inputStreamA: c.TermName, inputStreamB: c.TermName)(elementData: List[(c.universe.Type, c.universe.TermName, TreeOrderedBuf[c.type])]): c.Tree = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    elementData.foldLeft(Option.empty[Tree]) {
+      case (existingTreeOpt, (tpe, accessorSymbol, tBuf)) =>
+        existingTreeOpt match {
+          case Some(t) =>
+            val lastCmp = freshT("lastCmp")
+            Some(q"""
+              val $lastCmp = $t
+              if($lastCmp != 0) {
+                $lastCmp
+              } else {
+                ${tBuf.compareBinary(inputStreamA, inputStreamB)}
+              }
+              """)
+          case None =>
+            Some(tBuf.compareBinary(inputStreamA, inputStreamB))
+        }
+    }.getOrElse(q"0")
+  }
+
+  def hash(c: Context)(element: c.TermName)(elementData: List[(c.universe.Type, c.universe.TermName, TreeOrderedBuf[c.type])]): c.Tree = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    val currentHash = freshT("last")
+
+    val hashUpdates = elementData.map {
+      case (tpe, accessorSymbol, tBuf) =>
+        val target = freshT("target")
+        q"""
+          val $target = $element.$accessorSymbol
+          _root_.com.twitter.scalding.serialization.MurmerHashUtils.mixH1($currentHash, ${tBuf.hash(target)})
+          """
+    }
+
+    q"""
+      var $currentHash: Int = _root_.com.twitter.scalding.serialization.MurmerHashUtils.seed
+      ..${hashUpdates}
+      _root_.com.twitter.scalding.serialization.MurmerHashUtils.fmix($currentHash, ${elementData.size})
+    """
+  }
+
+  def put(c: Context)(inputStream: c.TermName, element: c.TermName)(elementData: List[(c.universe.Type, c.universe.TermName, TreeOrderedBuf[c.type])]): c.Tree = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+    val innerElement = freshT("innerElement")
+
+    elementData.foldLeft(q"") {
+      case (existingTree, (tpe, accessorSymbol, tBuf)) =>
+        q"""
+          $existingTree
+          val $innerElement = $element.$accessorSymbol
+          ${tBuf.put(inputStream, innerElement)}
+          """
+    }
+  }
+
+  def length(c: Context)(element: c.Tree)(elementData: List[(c.universe.Type, c.universe.TermName, TreeOrderedBuf[c.type])]): CompileTimeLengthTypes[c.type] = {
+    import c.universe._
+    import CompileTimeLengthTypes._
+    val (constSize, dynamicFunctions, maybeLength, noLength) =
+      elementData.foldLeft((0, List[c.Tree](), List[c.Tree](), 0)) {
```

Comment:
vector has fast append. List has very slow (O(N) vs O(log_32 N)). The difference might change compile times for wide products.